contractcourt/channel_arbitrator.go

```diff
@@ -1269,15 +1321,164 @@ func (c *ChannelArbitrator) isPreimageAvailable(hash lntypes.Hash) (bool,
+			}
+		}
+	}
+
+	// With both sets constructed, we'll now compute the set difference of
+	// our two sets of HTLCs. This'll give us the HTLCs that exist on the
+	// remote commitment transaction, but not on ours.
+	for htlcIndex, htlc := range remoteHTLCs {
+		if _, ok := localHTLCs[htlcIndex]; ok {
+			continue
+		}
+
+		pendingRemoteHTLCs = append(pendingRemoteHTLCs, htlc)
+	}
+
+	// Finally, we'll examine all the pending remote HTLCs for those that
+	// have expired. If we find any, then we'll recommend that they be
+	// failed now so we can free up the incoming HTLC.
+	for _, htlc := range pendingRemoteHTLCs {
+		// We'll now check if we need to go to chain in order to cancel
+		// the incoming HTLC.
+		goToChain := c.shouldGoOnChain(
+			htlc.RefundTimeout, c.cfg.OutgoingBroadcastDelta,
+			height,
+		)
+
+		// If we don't need to go to chain, and no commitments have
+		// been confirmed, then we can move on. Otherwise, if
+		// commitments have been confirmed, then we need to cancel back
+		// *all* of the pending remote HTLCS.
+		if !goToChain && !commitsConfirmed {
+			continue
+		}
+
+		log.Tracef("ChannelArbitrator(%v): immediately "+
+			"htlc=%x about to timeout on remote commitment",
+			c.cfg.ChanPoint, htlc.RHash[:])
+
+		actionMap[HtlcFailNowAction] = append(
+			actionMap[HtlcFailNowAction], htlc,
+		)
+	}
+
+	return actionMap
+}
+
 // prepContractResolutions is called either int he case that we decide we need
 // to go to chain, or the remote party goes to chain. Given a set of actions we
 // need to take for each HTLC, this method will return a set of contract
 // resolvers that will resolve the contracts on-chain if needed, and also a set
 // of packets to send to the htlcswitch in order to ensure all incoming HTLC's
 // are properly resolved.
-func (c *ChannelArbitrator) prepContractResolutions(htlcActions ChainActionMap,
+func (c *ChannelArbitrator) prepContractResolutions(
 	contractResolutions *ContractResolutions, height uint32,
-) ([]ContractResolver, []ResolutionMsg, error) {
+	trigger transitionTrigger,
+	confCommitSet *CommitSet) ([]ContractResolver, []ResolutionMsg, error) {
+
+	htlcSets := confCommitSet.toActiveHTLCSets()
+
+	// First, we'll reconstruct a fresh set of chain actions as the set of
+	// actions we need to act on may differ based on if it was our
+	// commitment, or they're commitment that hit the chain.
+	var (
+		htlcActions ChainActionMap
+		err         error
+	)
+	switch *confCommitSet.ConfCommitKey {
+
+	// If the local commitment transaction confirmed, then we'll examine
+	// that as well as their commitments to the set of chain actions.
+	case LocalHtlcSet:
+		htlcActions, err = c.checkLocalChainActions(
+			height, trigger, htlcSets, true,
+		)
+
+	// If the remote commitment confirmed, then we only need to examine the
+	// remote set for any HTLCs.
+	case RemoteHtlcSet:
+		htlcActions, err = c.checkCommitChainActions(
+			height, trigger, htlcSets[RemoteHtlcSet],
+		)
```

Comment:
That would be the case if there's no incoming HTLC at all, so it's a direct payment. In that case we could go to chain (as written in order to fully resolve it). However since we have an outgoing broadcast delta of zero atm, we would've went to chain anyway in order to resolve it.

Reply:
I am thinking about the case where we've sent commitSig only for a forwarded htlc. That htlc is then in the remote pending commit tx only. Then the remote party force closes the channel with the remote commit (not including that htlc). What makes us cancel the incoming htlc in that case?

Reply:
Ahh I see what you're getting at! Updated the tests to cover: "remote commit conf, with remote pending htlc" (htlc expired and not).

Reply:
And how about the other way around, remote pending confirms, with htlc still on the remote (htlc already failed on the remote pending)?

Reply:
If I'm following then then that's not possible. The only way the remote pending commit can have something that a remote commit doesn't is if it was manually failed back, and the only party that can fail back HTLCs is the one that received the HTLC.

Reply:
If the remote pending confirms, then we just play that as is. If the remote pending has the HTLC failed, then the remote party would've needed to send us a sig, then we revoke, meaning that we also don't have the HTLC locally on our commitment transaction.

Reply:
Isn't it in that case that we only cancel back our incoming htlc once we receive the revoke from the remote party? So:
* Remote party sends us a sig over a commit tx that doesn't contain the htlc anymore
* We send the revoke the to remote party
* We remove htlc from our remote tx and send a sig (it is now removed from remote pending)
* The remote party publishes that remote pending tx without sending a revoke
* What mechanism now lets us now cancel back the incoming?

Reply:
The only tx where the htlc is still in is the remote tx, but we don't inspect that

Reply:
Extended the tests to cover this case.